### PR TITLE
 Added support for profile field mentors

### PR DESCRIPTION
**Pull request by johnpanos**

### CODE REVIEW

File: classes/messenger/factories/course_recipient_send/message_recipient_send_factory.php [hunk 1]
```diff
@@ -128,12 +128,17 @@ private function send_message_to_mentor_user($mentor_user, $mentee_user)
...
         $mentor_users = $this->get_recipient_mentors();
+        $prepended_mentor_users = $this->get_recipient_prepended_mentors();
 
         $mentee_user = $this->recipient->get_user();
 
         foreach ($mentor_users as $mentor_user) {
             $this->send_message_to_mentor_user($mentor_user, $mentee_user);
         }
+
+        foreach($prepended_mentor_users as $mentor_user) {
```
Comment (johnpanos) [inline classes/messenger/factories/course_recipient_send/message_recipient_send_factory.php]: `send_message_to_mentor_user` won't work for fake accounts. Would be better to use `send_email_to_mentor_user` for profile field mentors when sending as a moodle message.

**Comment by vcs-edtech** [comment]
Any progress on adding this functionality and releasing it?

**Comment by chadmazilly** [comment]
I think it is a good contribution, but I'm a little hesitant to include it. My reasoning: the Moodle docs explain how to [create a mentor](https://docs.moodle.org/35/en/Parent_role) for a student. This block assumes that your system does it this way. Further, it looks like there is a [Mentees block](https://docs.moodle.org/35/en/Mentees_block) which also assumes this method. So if there is a standard way, why put in functionality that goes against that flow? Thoughts?

**Comment by vcs-edtech** [comment]
What we just need is the capability of sending email to additional email addresses in a user profile… currently a user can only receive an email to the one email address in their profile.

In our environment, we would add parent emails as additional email addresses in our users’ profiles.

In other environments, users could add their own second personal email address, a mentor, a tutor, a counselor, etc.

The mentor/parent role and mentor block - which are the same - do not provide for this capability.


Do you think the additional emails feature is something we can create together, or, should we move in a different direction?

Jeremy Cunningham, M.Ed
Education Technology Coordinator
VCS.net<http://vcs.net/> | @vcschools<http://twitter.com/vcschools> | @vcs_edtech<https://twitter.com/vcs_edtech>
408.513.2594
[VCS_logo_1C]

**Comment by chadmazilly** [comment]
This is a tool for everyone to use and to make better. I'll test and merge in tomorrow.

**Comment by rrusso** [comment]
We're implementing this feature with a few changes:

1. The admin settings for Quickmail will show a multi-select box for profile fields which the admin can select to send to the email addresses populated in the field instead of the prefix as specified in the pull-request.
2. If no profile fields are selected, the system will not bother to load the additional profile fields. We send millions of emails per semester and do not want/need any more overhead in this process. 
3. We will make sure any dynamic text fields math that of the associated user with an indication to the fact and NOT the fake user. Example: Hello [:firstname:] = Hello Robert (intended recipient).
4. We'll run a check to see if any of those fields were deleted from Moodle and remove them from the list of selected additional profile fields.
5. We're also looking at maybe periodically checking the emails for proper format and removing those that are invalid (not conforming to email address standards as defined in RFC 5322)

**Comment by johnpanos** [comment]
Would you have an estimate as to when your changes will be implemented?

**Comment by rrusso** [comment]
Chad is finishing up grade range notifications and once I've tested that and merged that into Master, this will be completed. It's only a few hours of work, so at the latest next week sometime. We're also adding a course entered event that's already been tested but waiting on grade range to merge all that in. We also added a new landing page for the "quickmail" breadcrumb that allows for tighter integration with more flexible themes. We plan on going live with everything on 10/07 in our prod environment so there's lots of testing involved and that adds to the overhead.

**Comment by chadmazilly** [comment]
Hey all, I've just pushed up some code to support all of this. Let me know what you think.

**Comment by johnpanos** [comment]
@chadmazilly We cloned the develop branch and took a look at it and it's great! This is exactly the feature we had wanted. Would you have any idea when the feature will be merged into master and released? Thanks for all your hard work on this so far, we really appreciate it.

**Comment by chadmazilly** [comment]
@johnpanos I just merged all into master as v2.1.0. Will work next week to get it up on the official Moodle plugin directory. Glad to hear it worked out for you!

**Pull request closed without merging**